File: day05.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define INPUT "day05.input"

int part1(int numbers[1069], int part2)
{
    int current, steps, move;
    current = steps = 0;

    while (current <= 1069 && current >= 0) {
        move = numbers[current];
        if (part2 && move >= 3)
            numbers[current] -= 1;
        else
            numbers[current] += 1;
        current += move;
        steps += 1;
    }

    return steps;
}

int main()
{
    FILE *fp;
    char *line = NULL;
    size_t len = 0;

    fp = fopen(INPUT, "r");
    if (fp == NULL) {
        perror(INPUT);
        exit(1);
    }

    int i = 0;
    int numbers1[1069];
    int numbers2[1069];

    while (getline(&line, &len, fp) != -1) {
        line[strcspn(line, "\r\n")] = 0;
        numbers1[i] = atoi(line);
        numbers2[i] = atoi(line);
        i++;
    }

    printf("%d\n", part1(numbers1, 0));
    printf("%d\n", part1(numbers2, 1));

    return 0;
}

File: day05.py
#!/usr/bin/env python

with open('day05.input') as f:
    data1 = [int(i) for i in f.read().splitlines()]
    data2 = list(data1)

def part1(part2, data):
    current = 0
    steps = 0

    while True:
        try:
            move = data[current]
            if part2 and move >= 3:
                data[current] -= 1
            else:
                data[current] += 1
            current += move
            steps += 1
        except IndexError:
            break

    return steps

print(part1(0, data1))
print(part1(1, data2))
